source/automem/utils.d

```diff
@@ -1,55 +1,90 @@
 module automem.utils;
 
-import std.traits: isFunctionPointer, isDelegate;
+// This is a destroy() copied and modified from
+// druntime, to allow for destruction attribute inference
 
-template hasNoGcDestructor(T) {
+void destruct(T)(T obj) if (is(T == class)) {
+    (cast(_finalizeType!T)&rt_finalize)(cast(void*)obj);
+}
 
-    import std.traits: functionAttributes, FunctionAttribute, hasMember;
+void destruct(T)(T obj) if (is(T == interface)) {
+    destruct(cast(Object)obj);
+}
 
-    static if(!is(T == class))
-        enum hasNoGcDestructor = false;
-    else static if(!hasMember!(T, "__dtor"))
-        enum hasNoGcDestructor = true;
-    else
-        enum hasNoGcDestructor = functionAttributes!(typeof(T.__dtor)) & FunctionAttribute.nogc;
+void destruct(T)(ref T obj) if (is(T == struct)) {
+    static if (__traits(hasMember, T, "__xdtor") &&
+               __traits(isSame, T, __traits(parent, obj.__xdtor)))
+        obj.__xdtor;
+}
 
+void destruct(T : U[n], U, size_t n)(ref T obj) if (!is(T == struct)) {
+    foreach_reverse (ref e; obj[])
+        destruct(e);
+}
 
+void destruct(T)(ref T obj)
+if(!is(T == struct) && !is(T == class) && !is(T == interface) && !_isStaticArray!T) {
+    obj = T.init;
 }
 
-// enum hasNoGcDestructor(T) = is(T == class) &&
-//     (!hasMember!(T, "__dtor") || (functionAttributes!(typeof(T.__dtor)) & FunctionAttribute.nogc));
+private:
+
+template _isStaticArray(T : U[n], U, size_t n) {
+    enum bool _isStaticArray = true;
+}
 
-@("hasNoGcDestructor")
-@safe pure unittest {
-    static assert(hasNoGcDestructor!NoGc);
-    static assert(!hasNoGcDestructor!Gc);
+template _isStaticArray(T) {
+    enum bool _isStaticArray = false;
 }
 
+template _Seq(TList...) {
```

Comment:
Yeah, it is. I basically copy-pasted it from my druntime sandbox, which in turn was a copypaste from druntime master.